module/zfs/arc.c

```diff
@@ -5803,10 +5803,11 @@ arc_getbuf_func(zio_t *zio, const zbookmark_phys_t *zb, const blkptr_t *bp,
 	arc_buf_t **bufp = arg;
 
 	if (buf == NULL) {
+		ASSERT(zio == NULL || zio->io_error != 0);
 		*bufp = NULL;
 	} else {
```

Comment:
It looks like you accidentally dropped an ASSERT from the OpenZFS version which should be here.

```
ASSERT(zio == NULL || zio->io_error == 0);
```

There are also two additional ASSERTs which were lost and should be added to `dbuf_prefetch_indirect_done()`.